### PR TITLE
ERC721UserMintable -> Template 2

### PR DESCRIPTION
**Pull request by efecarranza**
# Changelog

ERC721UserMintable (Template 2):

Add new contract file.
Add new error messages.
Add new ABI.
Add new unit tests.
Add new integration tests.

### CODE REVIEW

File: integration-test/erc721usermintable.contract.test-integration.js [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    const secretId = process.env.INFURA_PROJECT_SECRET;
+    const IPFS = { IPFSProjectID: '', IPFSProjectSecret: '' };
+
+    account = new Auth({
+      privateKey,
+      projectId,
+      secretId,
+      rpcUrl,
+      chainId,
+      IPFS,
```
Comment (salimtb) [inline integration-test/erc721usermintable.contract.test-integration.js:53]: IPFS will be a separated module , with another auth, you can delete this line .
example: 
`const ipfs = new IPFS( ...... ) , ipfs,uploadFIle(...)`

File: integration-test/erc721usermintable.contract.test-integration.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        baseURI: 'URI',
+        maxSupply: 10,
+        price: '0.00001',
+        maxTokenRequest: 1,
+      },
+    });
+  });
+
+  afterAll(async () => {
+    await server.close();
```
Comment (salimtb) [inline integration-test/erc721usermintable.contract.test-integration.js:71]: you don't need to open and close the connection to ganach, we do it in a global scope in the file `setupTests.js`

File: src/lib/ContractTemplates/ERC721UserMintable.js [hunk 3]
```diff
@@ -0,0 +1,942 @@
...
+    if (!name) {
+      throw new Error(
+        errorLogger({
+          location: ERROR_LOG.location.ERC721UserMintable_deploy,
+          message: ERROR_LOG.message.no_name_supplied,
+        }),
+      );
+    }
+
+    if (symbol === undefined) {
```
Comment (salimtb) [inline src/lib/ContractTemplates/ERC721UserMintable.js:93]: what if `symbol === null ` ??? i think it's better if you use `!symbol`, i know it's a copy past but are we allowed `null` as value for symbol ?

File: src/lib/ContractTemplates/ERC721UserMintable.js [hunk 4]
```diff
@@ -0,0 +1,942 @@
...
+    return newOptions;
+  }
+
+  /**
+   * Deploy ERC721UserMintable Contract. Used by the SDK class
+   * @param {string} name Name of the contract
+   * @param {string} symbol Symbol of the contract
+   * @param {string} baseURI baseURI for the contract
+   * @param {string} maxSupply Maximum supply of tokens for the contract
+   * @param {string} maxSupply Price to mint one token
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721UserMintable.js:61]: typo => remove "maxSupply" (it should be price)

File: src/lib/ContractTemplates/ERC721UserMintable.js [hunk 5]
```diff
@@ -0,0 +1,942 @@
...
+   * (link to a JSON file describing the contract's metadata)
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721UserMintable.js:62]: missing "maxTokenRequest" JSDoc info

File: src/lib/ContractTemplates/ERC721UserMintable.js [hunk 6]
```diff
@@ -0,0 +1,942 @@
...
+          message: ERROR_LOG.message.an_error_occured,
+          options: `${type} ${message}`,
+        }),
+      );
+    }
+  }
+
+  /**
+   * Sets the price (in Ether) of the mint
+   * @param price Price of the mint (per token) in Ether as a string
```
Comment (kalote) [inline src/lib/ContractTemplates/ERC721UserMintable.js:605]: missing type for param

**Review by kalote — CHANGES_REQUESTED** [review]
- Check the JSDoc of the ERC721UserMintable class
- there is some change to do on the SDK load & deploy method, no?
- Missing usage.js demonstration

**Comment by efecarranza** [comment]
@kalote updated. i didn't add an example in `usage.js` however. 
i don't think we need to do an example of each in that file, rather, it's a guide on how to set things up and how the SDK works.
we have the docs to add all the available templates.

**Review by salimtb — APPROVED** [review]

**Review by kalote — APPROVED** [review]

**Pull request merged**